File: src/erlang/ch01/1.3/l2/src/crt.erl
-module(crt).
-export([crt_pair/2, crt/1]).

%% crt_pair({A1,M1}, {A2,M2}) with coprime moduli -> {X, M1*M2}
crt_pair({A1,M1}, {A2,M2}) when M1 =< 0; M2 =< 0 ->
    erlang:error(badarg);
crt_pair({A1,M1}, {A2,M2}) ->
    %% Require gcd(M1,M2)=1
    {G, _, _} = extgcd_inv:extended_gcd(M1, M2),
    case (G =:= 1) orelse (G =:= -1) of
        false -> erlang:error({non_coprime_moduli, {M1, M2}});
        true  ->
            Inv = extgcd_inv:invmod(M1 rem M2, M2),
            K   = (((A2 - A1) rem M2) + M2) rem M2,
            K1  = (K * Inv) rem M2,
            M   = M1 * M2,
            X   = ((A1 + K1 * M1) rem M + M) rem M,
            {X, M}
    end.

%% Fold many congruences with pairwise coprime moduli
crt([]) -> erlang:error(badarg);
crt([{A,M}|Rest]) ->
    lists:foldl(fun({A2,M2}, {X,M1}) -> crt_pair({X,M1}, {A2,M2}) end,
                {A,M}, Rest).

File: archived/src/python/ch01/1.3/l2/crt.py
from typing import Iterable, Tuple, Optional
from extgcd_inv import invmod

def crt_pair(a1: int, m1: int, a2: int, m2: int) -> Tuple[int, int]:
    """
    Solve x ≡ a1 (mod m1), x ≡ a2 (mod m2) for coprime m1,m2.
    Return (x, m1*m2) with x in [0, m1*m2-1].
    Raise ValueError if moduli not coprime or inconsistent.
    """
    from math import gcd
    if gcd(m1, m2) != 1:
        raise ValueError("moduli must be coprime for crt_pair")
    k = ((a2 - a1) % m2) * invmod(m1 % m2, m2) % m2
    x = (a1 + k * m1) % (m1 * m2)
    return x, m1 * m2

def crt(congruences: Iterable[Tuple[int, int]]) -> Tuple[int, int]:
    """
    Fold CRT across many congruences with pairwise coprime moduli.
    Returns (x, M) s.t. x ≡ ai (mod mi) for all (ai, mi).
    """
    it = iter(congruences)
    try:
        a, m = next(it)
    except StopIteration:
        raise ValueError("no congruences")
    for a2, m2 in it:
        a, m = crt_pair(a, m, a2, m2)
    return a, m